docs: Cluster Peering for OSS Technical Preview (#13433)
hashicorp/consul: website/content/api-docs/peering.mdx

## Generate a Peering Token

This endpoint generates a peering token.

| Method  | Path                 | Produces           |
| ------- | -------------------- | ------------------ |
| `POST`  | `/peering/token`     | `application/json` |

The table below shows this endpoint's support for
[blocking queries](/api-docs/features/blocking),
[consistency modes](/api-docs/features/consistency),
[agent caching](/api-docs/features/caching), and
[required ACLs](/api#authentication).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required     |
| ---------------- | ----------------- | ------------- | ---------------- |
| `NO`             | `none`            | `none`        | `none`           |

### JSON Request Body Schema

- `PeerName` `(string: <required>)` - The name assigned to the peer cluster.
  The `PeerName` is used to reference the peer cluster in service discovery queries
  and configuration entries such as `service-intentions`. This field must be a
  valid DNS hostname label.

- `Partition` `(string: "")` - <EnterpriseAlert inline /> The admin partition that the
  peering token is generated from. Uses `default` when not specified.

> - `Datacenter` `(string: "")` - Specifies the datacenter. Defaults to the

Reviewer: ```suggestion
- `Datacenter` `(string: "")` - Specifies the datacenter. Defaults to the
```
Getting a little lost - this is the datacenter that you want to create peer connection with, correct? Can we say "target datacenter" or "external datacenter"? "Specifies the target datacenter for the peer connection."?

Reviewer 2: This is the datacenter where the token is generated. The datacenter that the `Partition` is in.

  agent's datacenter when not specified.

- `Token` `(string: "")` - Specifies the ACL token to use in the request. Takes precedence
  over the token specified in the `token` query parameter, `X-Consul-Token` request header,
  and `CONSUL_HTTP_TOKEN` environment variable.

- `Meta` `(map<string|string>: <optional>)` - Specifies KV metadata to associate with
  the peering. This parameter is not required and does not directly impact the cluster
  peering process.

### Sample Payload

```json
{
  "PeerName": "cluster-02",
  "Meta": {
    "env": "production"
  }
}
```

### Sample Request

```shell-session
$ curl --request POST \
   --header "X-Consul-Token: 5cdcae6c-0cce-4210-86fe-5dff3b984a6e" \
   --data @payload.json \
   http://127.0.0.1:8500/v1/peering/token
```

### Sample Response

```json
{
    "PeeringToken": "eyJDQSI6bnVsbCwiU2V..."
}
```

> ## Establish a Peering
> 
> This endpoint establishes a peering with a given peering token.

Reviewer: ```suggestion
## Establish a Peering Connection

This endpoint establishes a peering connection with a given peering token.
```

| `POST`  | `/peering/establish` | `application/json` |

This endpoint returns no data. Success or failure is indicated by the status
code returned.

The table below shows this endpoint's support for
[blocking queries](/api-docs/features/blocking),
[consistency modes](/api-docs/features/consistency),
[agent caching](/api-docs/features/caching), and
[required ACLs](/api#authentication).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required     |
| ---------------- | ----------------- | ------------- | ---------------- |
| `NO`             | `none`            | `none`        | `none`           |

### JSON Request Body Schema

- `PeerName` `(string: <required>)` - The name assigned to the peer cluster.
  The `PeerName` is used to reference the peer cluster in service discovery queries
  and configuration entries such as `service-intentions`. This field must be a
  valid DNS hostname label.

- `Partition` `(string: "")` - <EnterpriseAlert inline /> The admin partition
  that peers to the cluster that generated the peering token. Uses `default`
  when not specified.

- `PeeringToken` `(string: <required>)` - The peering token fetched from the
  peer cluster.

> - `Datacenter` `(string: "")` - Specifies the datacenter. Defaults to the
>   agent's datacenter when not specified.

Reviewer: ```suggestion
- `Datacenter` `(string: "")` - Specifies the datacenter. Defaults to the
  agent's datacenter when not specified.
```
Same as previous comment

Reviewer 2: Same as above, this is the datacenter where the peering is being established from. The datacenter of the `Partition`.

- `Token` `(string: "")` - Specifies the ACL token to use in the request. Takes precedence
  over the token specified in the `token` query parameter, `X-Consul-Token` request header,
  and `CONSUL_HTTP_TOKEN` environment variable.

- `Meta` `(map<string|string>: <optional>)` - Specifies KV metadata to associate with
  the peering. This parameter is not required and does not directly impact the cluster
  peering process.

### Sample Payload

```json
{
  "PeerName": "cluster-01",
  "PeeringToken": "eyJDQSI6bnVsbCwiU2V...",
  "Meta": {
    "env": "production"
  }
}
```

### Sample Request

```shell-session
$ curl --request POST \
   --header "X-Consul-Token: 5cdcae6c-0cce-4210-86fe-5dff3b984a6e" \
   --data @payload.json \
   http://127.0.0.1:8500/v1/peering/establish
```

### Sample Response

```json
{}
```

> ## Read a Peering
> 
> This endpoint reads a peering with the given name.

Reviewer: ```suggestion
## Read a Peering Connection

This endpoint returns information about a peering connection for the specified peer name.
```
Trying to use 'peering' as a description of the connection rather than the object, itself, to keep the language neutral.

[blocking queries](/api-docs/features/blocking),
[consistency modes](/api-docs/features/consistency),
[agent caching](/api-docs/features/caching), and
[required ACLs](/api#authentication).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `NO`             | `consistent`      | `none`        | `none`       |

### Path Parameters

- `name` `(string: <required>)` - Specifies the peering to read.

### Query Parameters

- `partition` `(string: "")` <EnterpriseAlert inline /> - Specifies the partition of the peering
  to read. If not specified will default to `default`.

### Sample Request

```shell-session
$ curl --header "X-Consul-Token: b23b3cad-5ea1-4413-919e-c76884b9ad60" \
   http://127.0.0.1:8500/v1/peering/cluster-02
```

### Sample Response

```json
{
    "ID": "462c45e8-018e-f19d-85eb-1fc1bcc2ef12",
    "Name": "cluster-02",
    "State": "INITIAL",
    "PeerID": "e83a315c-027e-bcb1-7c0c-a46650904a05",
    "PeerServerName": "server.dc1.consul",
    "PeerServerAddresses": [
        "10.0.0.1:8300"
    ],
    "CreateIndex": 89,
    "ModifyIndex": 89
}
```

> ## Delete a Peering
> 
> This endpoint marks a peering for deletion. Once marked, Consul will
> delete all the data imported from the peer in the background. Only once
> all associated data has been deleted will the peering actually disappear.
> Until then, further reads can be performed on the peering and a `DeletedAt`
> field will be populated with the timestamp of when the peering was
> marked for deletion.

Reviewer: ```suggestion
## Delete a Peering Connection

Call this endpoint to delete a peering connection. Consul deletes all data imported from the peer in the background. The peering connection is removed after all associated data has been deleted.
Operators can still read the peering connections while the data is being removed. A `DeletedAt`
field will be populated with the timestamp of when the peering was marked for deletion.
```
Is this cleaner?
